R/utils.R

```diff
@@ -340,3 +340,24 @@ flatten_dots <- function(...) {
   args <- rlang::dots_list(!!!args, .homonyms = "error")
   args
 }
+
+intersect_with_attr <- function(.l, ...) {
+  list <- c(.l, list(...))
+  attr_list <- purrr::map(list, attr_list)
+
+  intersect <- purrr::reduce(attr_list, intersect)
+  indices <- match(intersect, attr_list[[1]])
+  list[[1]][indices]
+}
+
+attr_list <- function(x) {
+  strip_attr <- function(x) {
+    attributes(x) <- NULL
+    x
+  }
+
+  list <- purrr::pmap(
+    c(list(strip_attr(x)), attributes(x)),
+    list
+  )
+}
```

Comment:
```suggestion
merge_attr_list <- function(x) {
  x_attr <- attributes(x)
  attributes(x) <- NULL
  # vector attributes are merged element-wise into the original list
  # scalar attributes are added to every element
  purrr::pmap(c(list(x), x_attr), list)
}
```

The `pmap(x, list)` idea is very clever as a way to merge a vector attribute into each list item. This works great when the attributes are all either length 1 or `length(x)`, but will likely choke for other lengths — `pmap()` will throw a bad length error. Do we need to worry about that?